theano/sandbox/cuda/acdc.py

```diff
@@ -0,0 +1,309 @@
+        return theano.Apply(self, [x_, A_, Ab_, D_, Db_], [x_.type()])
+
+    def grad(self, inputs, cost_grad):
+        inp, A, Ab, D, Db = inputs
+        gradOutput, = cost_grad
+        gradOutput = gpu_contiguous(gradOutput)
+        return FastACDCGrad()(inp, gradOutput, A, Ab, D, Db)
+
+    def c_support_code_apply(self, node, nodename):
+        # Based on a similar mechanism in sandbox/cuda/blas.py
+        #
+        # REMEMBER TO RAISE c_code_cache_version when changing any of
+        # these files
+        files = ['acdc.cu']
+        codes = [open(os.path.join(os.path.split(__file__)[0], f)).read()
+                for f in files]
+        return reduce(str.__add__, codes)
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        x, A, Ab, D, Db = inputs
+        z, = outputs
+        fail = sub['fail']
+        return """
+{
+    // Validate that the output storage exists.
+    if (NULL == %(z)s ||
+        CudaNdarray_NDIM(%(z)s) != CudaNdarray_NDIM(%(x)s))
+    {
+        /* Reference received to invalid output variable.
+        Decrease received reference's ref count and allocate new
+        output variable */
+        Py_XDECREF(%(z)s);
+        %(z)s = (CudaNdarray*)CudaNdarray_NewDims(CudaNdarray_NDIM(%(x)s),
+                                                  CudaNdarray_HOST_DIMS(%(x)s));
+    }
+
+    int batch_size;
+    int input_size;
+    int group_size;
+
+    if (CudaNdarray_NDIM(%(x)s) == 1) {
+        batch_size = 1;
+        group_size = 1;
+        input_size = CudaNdarray_HOST_DIMS(%(x)s)[0];
+    }
+    else if (CudaNdarray_NDIM(%(x)s) == 2) {
+        batch_size = CudaNdarray_HOST_DIMS(%(x)s)[0];
+        group_size = 1;
+        input_size = CudaNdarray_HOST_DIMS(%(x)s)[1];
+    }
+    else if (CudaNdarray_NDIM(%(x)s) == 3) {
+        batch_size = CudaNdarray_HOST_DIMS(%(x)s)[0];
+        group_size = CudaNdarray_HOST_DIMS(%(x)s)[1];
+        input_size = CudaNdarray_HOST_DIMS(%(x)s)[2];
+    }
+    else {
+        PyErr_Format(PyExc_ValueError, "Input must have 1, 2, or 3 dims");
+        %(fail)s;
+    }
+
+    int tmp_len = batch_size * input_size * group_size * 2;
+
+    // Declare auxiliary arrays.
+    CudaNdarray* tmp1 = (CudaNdarray*)CudaNdarray_NewDims(1, &tmp_len);
+    CudaNdarray* tmp2 = (CudaNdarray*)CudaNdarray_NewDims(1, &tmp_len);
+
+    // Verify.
+    if (!%(x)s || !%(z)s || !%(A)s || !%(Ab)s || !%(D)s || !%(Db)s || !tmp1 || !tmp2) {
+        PyErr_Format(PyExc_ValueError, "Could not allocate storage");
+        %(fail)s;
+    }
+
+    // cudaStream_t stream = THCState_getCurrentStream(state);
+    cudaStream_t stream = 0; // Assign the default stream?
+
+    // It should be up to the calling framework to provide plans.
+    // This is a hack to make it work, though will be very slow if the
+    // length/batchSize/groupSize change frequently.
+    static cufftHandle planR2C;
+    static cufftHandle planC2C;
+    static int planLength = -1;
+    static int planBatchSize = -1;
```

Comment:
Those will be shared by each op instance that get compiled. I think we want one cache per op instance. This would allow to not recompute it if we change the batch size for the test and train. This can be done by defining those methods instead of those static declaration:

```
    def c_support_code_struct(self, node, name):
        return """
cufftHandle planR2C%(name)s;
cufftHandle planC2C%(name)s;
int planLength%(name)s = -1;
int planBatchSize%(name)s = -1;
""" % dict(name=name)

    def c_init_code_struct(self, node, name, sub):
        return """
cufftHandle planR2C%(name)s = NULL;
cufftHandle planC2C%(name)s = NULL;
""" % dict(name=name)

    def c_cleanup_code_struct(self, node, name):
        return """
          cufftDestroy%(name)s(planR2C);
          cufftDestroy%(name)s(planC2C);
""" % dict(name=name)
```